### PR TITLE
Add transformation layer for converting ion response into UI oriented result.

### PR DESCRIPTION
**Pull request by haishengwu-okta**

### CODE REVIEW

File: test/unit/helpers/xhr/v2/FACTOR_VERIFICATION_REQUIRED_OKTA_PUSH.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        }
+      ]
+    },
+    "factor": {
+      "type": "object",
+      "value": {
+        "factorType": "push",
+        "provider": "okta",
+        "profile": {
+          "email": "omgitstom@abbott.dev"
```
Comment (nikhilvenkatraman-okta) [inline test/unit/helpers/xhr/v2/FACTOR_VERIFICATION_REQUIRED_OKTA_PUSH.js]: nit: use dummy email here

File: src/v2/ion/transformer.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  }
+
+  const firstLevelObjects = convertObjectType(resp);
+  const currentState = createCurrentStateObject(resp, Object.keys(firstLevelObjects));
+
+  const result = Object.assign({},
+    firstLevelObjects,
+    {
+      currentState,
+      __raw: resp,
```
Comment (nikhilvenkatraman-okta) [inline src/v2/ion/transformer.js]: nit: can we call this `__rawResponse`?

File: test/unit/spec/v2/ion/transformer_spec.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
+import transform from 'v2/ion/transformer';
+import XHRFactorRequiredEmail from '../../../helpers/xhr/v2/FACTOR_REQUIRED_EMAIL';
+import XHRFactorVerificationRequiredPush from '../../../helpers/xhr/v2/FACTOR_VERIFICATION_REQUIRED_OKTA_PUSH';
+
+describe('v2/ion/transformer', function () {
+  it('returns result when invokes with invalid resp', () => {
+    expect(transform()).toBeNull();
```
Comment (nikhilvenkatraman-okta) [inline test/unit/spec/v2/ion/transformer_spec.js]: maybe a future item might be to return some specific error when we cant parse the response.

File: test/unit/spec/v2/ion/transformer_spec.js [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+    expect(result.currentState['submit-factor']()).toEqual({
+      method: 'POST',
+      url: 'https://your-org.okta.com/api/v2/authn/',
+      data: {
+        stateHandle: '01OCl7uyAUC4CUqHsObI9bvFiq01cRFgbnpJQ1bz82'
+      }
+    });
+
+    expect(result.currentState['submit-factor']({foo: 'bar'})).toEqual({
```
Comment (nikhilvenkatraman-okta) [inline test/unit/spec/v2/ion/transformer_spec.js]: Would we ever call the function like this? where we are specifying  the data that needs to be posted? Wouldn't this be available in the response ?

Comment (haishengwu-okta) [reply test/unit/spec/v2/ion/transformer_spec.js]: sure, will  keep iterate

File: src/v2/ion/transformer.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+ *
+ * @returns {Object} actionObject
+ * @returns {string} actionObject.name
+ * @returns {function} actionObject.createRequest
+ */
+const createActionObj = (relObj) => {
+  const data = (relObj.value || [])
+    .filter(v => v.visible === false)
+    .map(v => { return { [v.name]: v.value }; })
+    .reduce((init, x) => Object.assign({}, init, x), {});
```
Comment (manueltanzi-okta) [inline src/v2/ion/transformer.js]: I think you can simplify to `.reduce((init, x) => Object.assign(init, x), {});` (might also be faster)

File: src/v2/ion/transformer.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+};
+
+/**
+ * @returns {Object.<string, function>} actions
+ */
+const createActions = (allCreateForms) => {
+  const allActions = allCreateForms.map(createActionObj);
+
+  // turns list of actions into a mapping
+  return allActions.reduce((init, action) => {
```
Comment (manueltanzi-okta) [inline src/v2/ion/transformer.js]: Also here, we might be able to simplify L100-105 to:
```suggestion
  return allActions.reduce((init, action) => Object.assign(init, { [action.name]: action.createRequest }), {});
```

File: src/v2/ion/transformer.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+};
+
+/**
+ *
+ * @param {AuthnResponse} resp authn v2 response
+ * @returns {AuthnResult} transformed result
+ */
+const convert = (resp) => {
+  if (!isObject(resp) || !resp) {
```
Comment (manueltanzi-okta) [inline src/v2/ion/transformer.js]: nit: Do we need the second statement? (I think `null` is not an object for `_`)

File: src/v2/ion/transformer.js [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+/**
+ *
+ * @returns {Object} actionObject
+ * @returns {string} actionObject.name
+ * @returns {function} actionObject.createRequest
+ */
+const createActionObj = (relObj) => {
+  const data = (relObj.value || [])
+    .filter(v => v.visible === false)
+    .map(v => { return { [v.name]: v.value }; })
```
[the remark this answers is not in the document]
Comment (haishengwu-okta) [reply src/v2/ion/transformer.js]: it probably will become `{ {[v.name]: v.value };  } which a little harder to read.

File: src/v2/ion/transformer.js (hunk 5 above)
Comment (haishengwu-okta) [reply src/v2/ion/transformer.js]: I kind of prefer not to override parameters (first one) unless it becomes issue or we have to.

File: src/v2/ion/transformer.js (hunk 6 above)
Comment (haishengwu-okta) [reply src/v2/ion/transformer.js]: same as last.

**Pull request merged**